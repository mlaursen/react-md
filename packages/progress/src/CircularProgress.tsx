import React, {
  CSSProperties,
  forwardRef,
  HTMLAttributes,
  useMemo,
} from "react";
import cn from "classnames";
import { bem } from "@react-md/utils";

import getProgress from "./getProgress";
import { ProgressProps } from "./types";

export interface CircularProgressProps
  extends Omit<HTMLAttributes<HTMLSpanElement>, "id">,
    ProgressProps {
  /**
   * An optional style to apply to the svg within the circular progress. The
   * values of this style object will be merged with the current determinate
   * style (if it exists).
   */
  svgStyle?: CSSProperties;

  /**
   * An optional className to apply to the svg within the circular progress.
   */
  svgClassName?: string;

  /**
   * An optional style to apply to the circle within the circular progress. The
   * values of this style object will be merged with the current determinate
   * style (if it exists).
   */
  circleStyle?: CSSProperties;

  /**
   * An optional className to apply to the circle within the cicular progress.
   */
  circleClassName?: string;

  /**
   * The radius for the circle. It is generally recommended to have the radius
   * be 1/2 of the viewbox and minus a few more pixels so that there is some
   * surrounding padding. You probably shouldn't really be changing this prop
   * though.
   */
  radius?: number;

  /**
   * The center point for the circle. This should be half of the `viewBox` prop
   * 99% of the time and probably won't be changed.
   */
  center?: number;

  /**
   * The viewbox for the child svg. I wouldn't recommend changing this value as
   * you will also need to update the `dashoffset` in both Sass and this prop to
   * get the animation to look nice again.
   */
  viewBox?: string;

  /**
   * The `stroke-dashoffset` for the circle within the SVG. You probably won't
   * be changing this value that much as it should match the
   * `$rmd-progress-circle-dashoffset` Sass variable. This is really just used
   * to help to create the determinite progress animation.
   */
  dashoffset?: number;

  /**
   * The max rotation value for the circular progress. If you set this value to
   * a number less than or equal to 0, the circular progress will no longer
   * rotate with the determinate progress type.
   */
  maxRotation?: number;

  /**
   * Boolean if the circular progress should be centered using left and right
   * margins.
   */
  centered?: boolean;

  /**
   * Boolean if the smaller size should be used instead.
   *
   * @since 2.3.0
   */
  small?: boolean;
}

const block = bem("rmd-circular-progress");

const CircularProgress = forwardRef<HTMLSpanElement, CircularProgressProps>(
  function CircularProgress(
    {
      className,
      svgStyle: propSvgStyle,
      svgClassName,
      circleStyle: propCircleStyle,
      circleClassName,
      value,
      min = 0,
      max = 100,
      radius = 30,
      center = 33,
      viewBox = "0 0 66 66",
      dashoffset = 187,
      animate = true,
      centered = true,
      maxRotation = 360 * 1.75,
      small = false,
      ...props
    },
    ref
  ) {
    const progress = getProgress(min, max, value);
    const svgStyle = useMemo<CSSProperties | undefined>(() => {
      if (typeof progress !== "number") {
        return propSvgStyle;
      }

      let transform = propSvgStyle && propSvgStyle.transform;
      if (maxRotation > 0) {
        const rotate = `rotate(${maxRotation * progress}deg)`;
        transform = `${rotate}${transform ? ` ${transform}` : ""}`;
      }

      return {
        ...propSvgStyle,
        WebkitTransform: transform,
        transform,
      };
    }, [progress, maxRotation, propSvgStyle]);

    const circleStyle = useMemo(() => {
      if (typeof progress !== "number") {
        return propCircleStyle;
      }

      return {
        ...propCircleStyle,
        strokeDashoffset: dashoffset - dashoffset * progress,
      };
    }, [progress, propCircleStyle, dashoffset]);

    const determinate = typeof progress === "number";
    const indeterminate = !determinate;
    return (
      <span
        {...props}
        ref={ref}
        role="progressbar"
        aria-valuemin={min}
        aria-valuemax={max}
        aria-valuenow={value}
        className={cn(block({ centered, small }), className)}
      >
        <svg
          style={svgStyle}
          className={cn(
            block("svg", {
              animate: animate && determinate,
              determinate,
              indeterminate,
            }),
            svgClassName
          )}
          viewBox={viewBox}
        >
          <circle
            style={circleStyle}
            className={cn(
              block("circle", {
                animate: animate && determinate,
                determinate,
                indeterminate,
              }),
              circleClassName
            )}
            r={radius}
            cx={center}
            cy={center}
          />
        </svg>
      </span>
    );
  }
);

if (process.env.NODE_ENV !== "production") {
  try {
    const PropTypes = require("prop-types");

    CircularProgress.propTypes = {
      id: PropTypes.string.isRequired,
      className: PropTypes.string,
      min: PropTypes.number,
      max: PropTypes.number,
      value: PropTypes.number,
      animate: PropTypes.bool,
      svgStyle: PropTypes.object,
      svgClassName: PropTypes.string,
      circleStyle: PropTypes.object,
      circleClassName: PropTypes.string,
      radius: PropTypes.number,
      center: PropTypes.number,
      centered: PropTypes.bool,
      maxRotation: PropTypes.number,
      dashoffset: PropTypes.number,
      small: PropTypes.bool,
      viewBox: PropTypes.string,
    };
  } catch (e) {}
}

export default CircularProgress;
