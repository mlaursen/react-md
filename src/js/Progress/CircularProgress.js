import React, { PureComponent, PropTypes } from 'react';
import cn from 'classnames';
import isBetween from '../utils/NumberUtils/isBetween';

const ROATE_DISTANCE = 360 * 1.75;
const BASE_SIZE = 24; // font-icon font size

/**
 * The `CircularProgress` component is used to give visual feedback while your app
 * is loading.
 *
 * There are two different types of circular progress bars: `Determinate` and `Indeterminate`.
 *
 * A `Determinate` circular progress bar should be used when you want to keep track of the current
 * progress. An example would be downloading a file.
 *
 * An `Indeterminate` circular progress bar should be used when you can not keep track of the progress
 * yourself. An example would be waiting for some API call to complete.
 */
export default class CircularProgress extends PureComponent {
  static propTypes = {
    /**
     * A style object to apply to the svg. If this is a determinate `CircularProgress`,
     * the `transform` (and vendor prefixes) styles will be merged with the current
     * progress rotation.
     */
    style: PropTypes.object.isRequired,

    /**
     * An optional className to apply to the svg.
     */
    className: PropTypes.string,

    /**
     * The current value of the progress. If this value is defined, it will
     * be converted to a determinate circular progress. The progress will not
     * advance unless this value changes.
     *
     * This value should also be a number between 0 and 100.
     */
    value: (props, propName, component, ...others) => {
      if (typeof props[propName] === 'undefined') { return null; }
      let err = PropTypes.number(props, propName, component, ...others);
      if (!err) {
        const value = props[propName];
        if (!isBetween(value, 0, 100)) {
          err = new Error(
            `A determinate '${component}' was given a value '${value}'. The 'value' prop should be between 0 and 100`
          );
        }
      }

      return err;
    },

    /**
     * The scale for the circular progress.
     */
    scale: PropTypes.number.isRequired,

    /**
     * You probably don't want to update this. I'm not good at svg. This should
     * match the scss variable `$md-circular-progress-stroke-dashoffset`.
     */
    determinateDashoffset: PropTypes.number.isRequired,

    /**
     * Boolean if the the progress should be centered in it's container.
     */
    centered: PropTypes.bool,
  };

  static defaultProps = {
    style: {},
    scale: 1,
    determinateDashoffset: 187,
    centered: true,
  };

  render() {
    const {
      scale,
      style,
      className,
      value,
      determinateDashoffset,
      centered,
      ...props,
    } = this.props;

    const isDeterminate = typeof value === 'number';
    let circleStyle;
    let svgStyle = style;
    if (isDeterminate) {
      const rotate = `rotate(${ROATE_DISTANCE / 100 * value}deg)`;
      circleStyle = {
        strokeDashoffset: determinateDashoffset - (determinateDashoffset / 100 * value),
      };

      svgStyle = Object.assign({}, style, {
        WebkitTransform: cn(style.WebkitTransform, rotate),
        MozTransform: cn(style.MozTransform, rotate),
        transform: cn(style.transform, rotate),
      });
    }

    return (
      <svg
        {...props}
        style={svgStyle}
        className={cn('md-circular-progress', className, {
          centered,
          'determinate': isDeterminate,
          'indeterminate': !isDeterminate,
        })}
        width={scale * BASE_SIZE}
        height={scale * BASE_SIZE}
        viewBox="0 0 66 66"
      >
        <circle
          className="md-circular-progress-path"
          strokeWidth="6"
          strokeLinecap="round"
          style={circleStyle}
          cx="33"
          cy="33"
          r="30"
        />
      </svg>
    );
  }
}
