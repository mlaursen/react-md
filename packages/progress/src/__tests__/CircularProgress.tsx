import React, { CSSProperties } from "react";
import { render } from "@testing-library/react";

import CircularProgress from "../CircularProgress";

describe("CircularProgress", () => {
  it("should render correctly", () => {
    const { container, rerender } = render(<CircularProgress id="progress" />);

    expect(container).toMatchSnapshot();

    rerender(<CircularProgress id="progress" centered />);
    expect(container).toMatchSnapshot();

    rerender(<CircularProgress id="progress" centered={false} />);
    expect(container).toMatchSnapshot();
  });

  it("should merge the transform style if it exists", () => {
    const style: CSSProperties = {
      WebkitTransform: "translateX(20px)",
      transform: "translateX(20px)",
    };

    const { getByRole } = render(
      <CircularProgress id="progress" style={style} value={20} />
    );
    const progress = getByRole("progressbar");
    expect(progress.style.transform).toContain(style.transform);
    expect(progress).toMatchSnapshot();
  });

  it("should support a small variant", () => {
    const { getByRole, rerender } = render(
      <CircularProgress id="small=pgoress" small />
    );
    const progress = getByRole("progressbar");

    expect(progress.className).toContain("rmd-circular-progress--small");
    expect(progress).toMatchSnapshot();

    rerender(
      <CircularProgress id="small-circular-progress" small centered={false} />
    );

    expect(progress.className).toContain("rmd-circular-progress--small");
    expect(progress).toMatchSnapshot();
  });
});
