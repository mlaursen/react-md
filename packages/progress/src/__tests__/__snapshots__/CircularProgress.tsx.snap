// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularProgress should merge the transform style if it exists 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="20"
  class="rmd-circular-progress rmd-circular-progress--centered"
  id="progress"
  role="progressbar"
  style="transform: translateX(20px);"
>
  <svg
    class="rmd-circular-progress__svg rmd-circular-progress__svg--animate rmd-circular-progress__svg--determinate"
    style="transform: rotate(126deg);"
    viewBox="0 0 66 66"
  >
    <circle
      class="rmd-circular-progress__circle rmd-circular-progress__circle--animate rmd-circular-progress__circle--determinate"
      cx="33"
      cy="33"
      r="30"
      style="stroke-dashoffset: 149.6;"
    />
  </svg>
</span>
`;

exports[`CircularProgress should render correctly 1`] = `
<div>
  <span
    aria-valuemax="100"
    aria-valuemin="0"
    class="rmd-circular-progress rmd-circular-progress--centered"
    id="progress"
    role="progressbar"
  >
    <svg
      class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
      viewBox="0 0 66 66"
    >
      <circle
        class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
        cx="33"
        cy="33"
        r="30"
      />
    </svg>
  </span>
</div>
`;

exports[`CircularProgress should render correctly 2`] = `
<div>
  <span
    aria-valuemax="100"
    aria-valuemin="0"
    class="rmd-circular-progress rmd-circular-progress--centered"
    id="progress"
    role="progressbar"
  >
    <svg
      class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
      viewBox="0 0 66 66"
    >
      <circle
        class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
        cx="33"
        cy="33"
        r="30"
      />
    </svg>
  </span>
</div>
`;

exports[`CircularProgress should render correctly 3`] = `
<div>
  <span
    aria-valuemax="100"
    aria-valuemin="0"
    class="rmd-circular-progress"
    id="progress"
    role="progressbar"
  >
    <svg
      class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
      viewBox="0 0 66 66"
    >
      <circle
        class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
        cx="33"
        cy="33"
        r="30"
      />
    </svg>
  </span>
</div>
`;

exports[`CircularProgress should support a small variant 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-circular-progress rmd-circular-progress--centered rmd-circular-progress--small"
  id="small=pgoress"
  role="progressbar"
>
  <svg
    class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
    viewBox="0 0 66 66"
  >
    <circle
      class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
      cx="33"
      cy="33"
      r="30"
    />
  </svg>
</span>
`;

exports[`CircularProgress should support a small variant 2`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-circular-progress rmd-circular-progress--small"
  id="small-circular-progress"
  role="progressbar"
>
  <svg
    class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
    viewBox="0 0 66 66"
  >
    <circle
      class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
      cx="33"
      cy="33"
      r="30"
    />
  </svg>
</span>
`;
